{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7792e95-d2f3-4a65-a996-559e0153c831",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''--------------------- Exception Handling Assignment 2-------------------------\n",
    "Q 1 - Explain why we have to use the Exception ------------------------\n",
    "Ans - We have to use the Exception class while creating the Custom Exception because it allows us to customize\n",
    "the exceptions as per our requirement. By deriving the custome exception class from Exception class, the custom \n",
    "exception class will get the flexibility to define our own behaviour and attributes for the exception.\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58da7620-a83a-4877-8ca6-b414f40204c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Q 2 - Write a python program to print --------------\n",
    "Ans -\n",
    "'''\n",
    "import inspect\n",
    "import logging\n",
    "\n",
    "logging.basicConfig(filename=\"test1.log\", level=logging.DEBUG, format=\"%(asctime)s %(message)s\")\n",
    "\n",
    "def treeClass(cls, ind=0):\n",
    "    logging.info(\"{} {}\".format('-' * ind, cls.__name__))\n",
    "    for i in cls.__subclasses__():\n",
    "        treeClass(i, ind+3)\n",
    "\n",
    "logging.info(\"Hierarchy of Python Exception is: -\") \n",
    "\n",
    "inspect.getclasstree(inspect.getmro(BaseException))\n",
    "treeClass(BaseException)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9adabb86-02ed-4b97-a38c-bd3936f59219",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Q 3 -What errors are defined in ArithmeticError ----------------\n",
    "Ans - \n",
    "Errors defined in the ArithmeticError class are: -\n",
    "1) FloatingPointError\n",
    "2) OverflowError\n",
    "3) ZeroDivisionError\n",
    "\n",
    "ZeroDivisionError - A ZeroDivisionError is raised when the second argument of a division operation is zero. \n",
    "This exception returns a string indicating the type of the operands and the operation.\n",
    "\n",
    "OverflowError - An OverflowError is raised when the result of an arithmetic operation is out of range.\n",
    "'''\n",
    "import logging\n",
    "logging.basicConfig(filename=\"test1.log\", level=logging.DEBUG, format=\"%(asctime)s %(message)s\")\n",
    "\n",
    "try:\n",
    "    x=10\n",
    "    y=int(input(\"Enter the number : \"))\n",
    "    z=0\n",
    "    z=x/y\n",
    "    logging.info(\"Result = {}\".format(z))\n",
    "except ZeroDivisionError as e:\n",
    "    logging.error(\"{} divided by zero.\".format(x))\n",
    "\n",
    "try:\n",
    "    var_x = math.exp(700)\n",
    "    logging.info(\"Control is Off : Result = {}\".format(var_x))\n",
    "    \n",
    "    var_y = math.exp(1000)\n",
    "    logging.info(\"Control is On : Result = {}\".format(var_y))\n",
    "    \n",
    "except OverflowError as e:\n",
    "    logging.error(\"Its a Floating Point Error.\")\n",
    "    \n",
    "logging.shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f45e10b-a54e-4a9d-b78f-baa642c4dbfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Q 4 - Why LookupError class is used-----------------------\n",
    "Ans - This is the base class for those exceptions that are raised when a key or index used on a mapping or sequence is invalid or not found. The exceptions raised are :\n",
    "KeyError - KeyError is raised when the mapping key is not found in the set of existing keys.\n",
    "IndexError - IndexError is raised when a sequence is referenced which is out of range.\n",
    "'''\n",
    "import logging\n",
    "logging.basicConfig(filename=\"test1.log\", level=logging.DEBUG, format=\"%(asctime)s %(message)s\")\n",
    "\n",
    "try: \n",
    "    a = [1, 2, 3] \n",
    "    logging.info(\"{}\".format(a[3]))\n",
    "\n",
    "except IndexError: \n",
    "    logging.error(\"Index out of bound error.\")\n",
    "    \n",
    "else: \n",
    "    logging.info(\"Success\")\n",
    "\n",
    "try:\n",
    "    dict_array = {'a':1, 'b':2, 'c':3,'d':4}\n",
    "    logging.info(\"{}\".format(dict_array['x']))\n",
    "    \n",
    "except KeyError:\n",
    "    logging.error(\"Key not found in dictionary\")\n",
    "\n",
    "else: \n",
    "    logging.info(\"Success\")\n",
    "\n",
    "logging.info(\"Execution completed.\")\n",
    "logging.shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59de5fe8-fbf2-4bdf-89c2-1f1832cfecfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Q 5 - Explain ImportError.----------------------\n",
    "Ans - \n",
    "ImportError - When we want to import an module in our program, we use import statement. When program is not able\n",
    "to import the module, it raises an exception called ImportError. It is also raised when the \"from list\" in the\n",
    "from..import has a name that cannot be found.\n",
    "Most probable reason for ImportError - \n",
    "a) module is not present at all \n",
    "b) name specified of module is wrong.\n",
    "\n",
    "ModuleNotFoundError - ModuleNotFoundError exception is raised when python is not able to locate the module. It\n",
    "is a subclass of ImportError.\n",
    "Reasons for this are as-\n",
    "a) Incorrect spelling of the module name\n",
    "b) module is not installed\n",
    "c) incorrect python version\n",
    "d) incorrect file path\n",
    "e) different python environment\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bac34352-0e96-4dba-b51c-3fd33524e8a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Q 6 - List down some best practices ---------------------\n",
    "Ans - Best practices for exception handling in python are as follows: -\n",
    "1) Always use a specific exception.\n",
    "2) Always specify appropriate message.\n",
    "3) Always try to logging error.\n",
    "4) Always avoid to write multiple exception handling.\n",
    "5) Document all the errors.\n",
    "6) Clean up all the resources.\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
